### PR TITLE
refactor: refactor resourceCreator to class

### PR DESCRIPTION
**Pull request by GuillaumeCisco**
## Pull Request checklist:

- [x] Write an explicit title for the Pull Request, following [Conventional Commits specification](https://www.conventionalcommits.org)
- [x] Create automatic tests
- [x] Test manually the implemented changes
- [x] Review my own code (indentation, syntax, style, simplicity, readability)
- [x] Wonder if you can improve the existing code

### CODE REVIEW

File: src/services/resource-creator.js [hunk 1]
```diff
@@ -5,55 +5,71 @@ const { ErrorHTTP422 } = require('./errors');
...
+  async handleSave(record, callback) {
+    const { associations } = this.model;
+    if (associations) {
+      const promisesBeforeSave = Object.entries(associations)
+        .reduce(callback(record, this.params), []);
+      await P.all(promisesBeforeSave);
+    }
+  }
+
+  async perform() {
```
Comment (slimee) [inline src/services/resource-creator.js:49]: The create operation should be a very readable and simple list of invocation for example:

perform:
  // buildInstance
  // handleAssociationsBeforeSave
  // saveInstance (validate then save)
  // handleAssociationsAfterSave
  // appendCompositePrimary
  // return makeResourceGetter()

File: src/services/resource-creator.js [hunk 2]
```diff
@@ -5,55 +5,71 @@ const { ErrorHTTP422 } = require('./errors');
...
+    setter = `set${_.upperFirst(name)}`;
+  } else if (['BelongsToMany', 'HasMany'].includes(associationType)) {
+    setter = `add${_.upperFirst(name)}`;
+  }
+  return setter;
+};
+
+const getPromisesBeforeSave = (record, params) => (promises, [name, association]) => {
+  if (association.associationType === 'BelongsTo') {
+    promises.push(record[`set${_.upperFirst(name)}`](params[name], { save: false }));
```
Comment (slimee) [inline src/services/resource-creator.js:20]: This deserve more lines to be more clear:

```
const getPromisesBeforeSave = (record, params) => (promises, [fieldName, association]) => {
  if (association.associationType === 'BelongsTo') {
    const methodName = `set${_.upperFirst(fieldName)}`;
    const method = record[methodName];
    const promise = method(params[fieldName], { save: false });
    promises.push(promise);
  }
  return promises;
};
```

File: src/services/resource-creator.js [hunk 3]
```diff
@@ -5,55 +5,71 @@ const { ErrorHTTP422 } = require('./errors');
...
-    const promises = [];
-    const recordCreated = model.build(params);
-
-    if (model.associations) {
-      _.forOwn(model.associations, (association, name) => {
-        if (association.associationType === 'BelongsTo') {
-          promises.push(recordCreated[`set${_.upperFirst(name)}`](params[name], { save: false }));
-        }
-      });
+const getSetter = (associationType, name) => {
```
Comment (slimee) [inline src/services/resource-creator.js:8]: Since it returns a string maybe it should be `getSetterName`

**Review by slimee — CHANGES_REQUESTED** [review]
yeah!

Some comments here but very nice job

File: src/services/resource-creator.js [hunk 4]
```diff
@@ -5,55 +5,71 @@ const { ErrorHTTP422 } = require('./errors');
...
+const getPromisesAfterSave = (record, params) => (promises, [name, association]) => {
+  const setter = getSetter(association.associationType, name);
+  if (setter) {
+    promises.push(record[setter](params[name]));
+  }
+  return promises;
+};
+
+class ResourceCreator {
+  constructor(model, params) {
```
[the remark this answers is not in the document]
Comment (slimee) [reply src/services/resource-creator.js:34]: Said in call: this signature is shared with forest-express.. Wont fix now.

[the remark this answers is not in the document]
Comment (GuillaumeCisco) [reply src/services/resource-creator.js:34]: As it is called by forest-express, I cannot modify it without modifying the code in forest-express :/

File: src/services/resource-creator.js [hunk 5]
```diff
@@ -5,55 +5,71 @@ const { ErrorHTTP422 } = require('./errors');
...
+    this.model = model;
+    this.params = params;
+    this.schema = Interface.Schemas.schemas[model.name];
+  }
+
+  async handleSave(record, callback) {
```
[the remark this answers is not in the document]
Comment (GuillaumeCisco) [reply src/services/resource-creator.js:47]: The linter will complain with:
```
Unexpected dangling '_' in '_getPromisesBeforeSave'  no-underscore-dangle
```
:/

[the remark this answers is not in the document]
Comment (slimee) [reply src/services/resource-creator.js:47]: I think you can add "no-underscore-dangle": 0 in .eslintrc
Are you ok with that @arnaudbesnier ?

File: src/services/resource-creator.js [hunk 6]
```diff
@@ -5,55 +5,85 @@ const { ErrorHTTP422 } = require('./errors');
...
+    const { associations } = this.model;
+    if (associations) {
+      callback = callback.bind(this);
+      const promisesBeforeSave = Object.entries(associations).reduce(callback(record), []);
+      await P.all(promisesBeforeSave);
+    }
+  }
+
+  async perform() {
+    // buildInstance
```
Comment (slimee) [inline src/services/resource-creator.js:57]: I was thinking of really creating the methods named buildInstance, handleAssociationsBeforeSave, saveInstance, etc.

File: src/services/resource-creator.js [hunk 7]
```diff
@@ -5,55 +5,85 @@ const { ErrorHTTP422 } = require('./errors');
...
-          promises.push(recordCreated[`set${_.upperFirst(name)}`](params[name], { save: false }));
-        }
-      });
+const getSetterName = (associationType, name) => {
+  let setterName;
+  if (associationType === 'HasOne') {
+    setterName = `set${_.upperFirst(name)}`;
+  } else if (['BelongsToMany', 'HasMany'].includes(associationType)) {
+    setterName = `add${_.upperFirst(name)}`;
+  }
```
Comment (slimee) [inline src/services/resource-creator.js:14]: I still feel unconfortable with unreachable code.
Maybe you can merge this inside `getPromisesAfterSave`

File: src/services/resource-creator.js [hunk 8]
```diff
@@ -5,55 +5,85 @@ const { ErrorHTTP422 } = require('./errors');
...
+};
+
+class ResourceCreator {
+  constructor(model, params) {
+    this.model = model;
+    this.params = params;
+    this.schema = Interface.Schemas.schemas[model.name];
+  }
+
+  getPromisesBeforeSave(record) {
```
Comment (slimee) [inline src/services/resource-creator.js:25]: this is a factory method and should be named like makeXXX

File: src/services/resource-creator.js [hunk 9]
```diff
@@ -5,55 +5,85 @@ const { ErrorHTTP422 } = require('./errors');
...
+      if (association.associationType === 'BelongsTo') {
+        const setterName = `set${_.upperFirst(name)}`;
+        const promise = record[setterName](this.params[name], { save: false });
+        promises.push(promise);
+      }
+      return promises;
+    };
+  }
+
+  getPromisesAfterSave(record) {
```
Comment (slimee) [inline src/services/resource-creator.js:36]: Same it's a factory method

File: src/services/resource-creator.js (hunk 7 above)
Comment (GuillaumeCisco) [reply src/services/resource-creator.js:14]: I will try :)
Hoping codeclimate is ok with that 🤞

File: src/services/resource-creator.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (arnaudbesnier) [reply src/services/resource-creator.js:47]: As we have this rule activated in the backend code base, yes I am ok with that.
Always better to have the same rules everywhere.

File: src/services/resource-creator.js (hunk 6 above)
Comment (GuillaumeCisco) [reply src/services/resource-creator.js:52]: There is only 16 lines of code, so I think it should be ok :)

**Review by slimee — APPROVED** [review]
🚀 

It's brave to start modernize this part, since the interface is shared with forest-ssequeoize. 💪

**Pull request merged**